File: web/src/lib/mainlandMembershipPurchase.ts
import type { MembershipTier } from '@/lib/api';

/** Mainland lifetime tier row visibility (matches Flutter `shouldShowMainlandTier`). */
export function shouldShowMainlandTier(tier: MembershipTier, currentTierCode: string): boolean {
  if (tier.productType === 'ADDON') return true;
  if (tier.code.toUpperCase() !== 'PRO') return false;
  return currentTierCode.toUpperCase() === 'FREE';
}

export function isMainlandTierPurchaseDisabled(
  tier: MembershipTier,
  currentTierCode: string,
  canBuyAddon: boolean,
  pendingOrder: boolean,
): boolean {
  if (pendingOrder) return true;
  if (tier.productType === 'ADDON') return !canBuyAddon;
  if (tier.code.toUpperCase() === 'PRO') {
    return currentTierCode.toUpperCase() !== 'FREE';
  }
  return true;
}

/** List price only — no Mini→Pro upgrade diff in UI. */
export function mainlandTierDisplayPriceCent(tier: MembershipTier): number {
  return tier.priceCent;
}

File: app/lib/utils/mainland_membership_purchase.dart
import '../api/membership.dart';

/// Whether a mainland (CNY buyout) tier row should appear in the membership list.
/// 新购仅 Pro + 增购；Mini 已停售，存量 Mini 用户不再展示 Pro 升级。
bool shouldShowMainlandTier(MembershipTier tier, String currentTierCode) {
  if (tier.isAddon) return true;
  if (tier.code.toUpperCase() != 'PRO') return false;
  return currentTierCode.toUpperCase() == 'FREE';
}

/// Mainland lifetime tier / add-on purchase button state.
bool isMainlandTierPurchaseDisabled(
  MembershipTier tier,
  MembershipMe? me, {
  required bool pendingOrder,
  required bool purchasing,
}) {
  if (pendingOrder || purchasing) return true;
  if (tier.isAddon) return !(me?.canBuyAddon ?? false);
  final current = (me?.tierCode ?? 'FREE').toUpperCase();
  if (tier.code.toUpperCase() == 'PRO') {
    return current != 'FREE';
  }
  return true;
}

/// Price shown on the card (list price; no tier upgrade diff in UI).
int mainlandTierDisplayPriceCent(MembershipTier tier) {
  return tier.priceCent;
}
